### ISSUE TITLE
Proposal for Code Structure Improvement Using `jax.lax.cond`

### ISSUE THREAD

**Issue by helpingstar** [instadeepai/jumanji#245]
### Is your feature request related to a problem? Please describe

This is a simple question regarding code style. It is not related to any bugs.

https://github.com/instadeepai/jumanji/blob/fd511b4523425bdcc69f2fd3c8869cefddd49f1d/jumanji/environments/routing/connector/env.py#L184-L198

https://github.com/instadeepai/jumanji/blob/fd511b4523425bdcc69f2fd3c8869cefddd49f1d/jumanji/environments/logic/game_2048/env.py#L222-L236

Rather than repeatedly using `lambda` and duplicating variables as shown in the code above, it seems better to follow the functional style of `jax.lax.cond` and write it in the style of the solution code below.

It seems that there is little to no difference in performance.
If this is a minor issue, I will close it.

### Describe the solution you'd like

```python
timestep = jax.lax.cond(
    done,
    termination,
    transition,
    reward,
    observation,
    extras,
)
```

### Describe alternatives you've considered
None

### Additional context

https://github.com/instadeepai/jumanji/blob/fd511b4523425bdcc69f2fd3c8869cefddd49f1d/jumanji/environments/logic/minesweeper/env.py#L178-L185

https://github.com/instadeepai/jumanji/blob/fd511b4523425bdcc69f2fd3c8869cefddd49f1d/jumanji/environments/logic/graph_coloring/env.py#L202-L209

https://github.com/instadeepai/jumanji/blob/fd511b4523425bdcc69f2fd3c8869cefddd49f1d/jumanji/environments/logic/rubiks_cube/env.py#L169-L176

https://github.com/instadeepai/jumanji/blob/fd511b4523425bdcc69f2fd3c8869cefddd49f1d/jumanji/environments/logic/sudoku/env.py#L124-L132

https://github.com/instadeepai/jumanji/blob/fd511b4523425bdcc69f2fd3c8869cefddd49f1d/jumanji/environments/packing/tetris/env.py#L235-L242

---

### Misc
- [X] Check for duplicate requests.

**Comment by sash-a** [comment]
Hi, thanks for the suggestion! Agreed it isn't as clean as the others, but the reason it needs to be done this way is because the `transition` and `termination` function take different arguments:

```python
def transition(
    reward: Array,
    observation: Observation,
    discount: Optional[Array] = None,
    extras: Optional[Dict] = None,
    shape: Union[int, Sequence[int]] = (),
) -> TimeStep:
```

```python
def termination(
    reward: Array,
    observation: Observation,
    extras: Optional[Dict] = None,
    shape: Union[int, Sequence[int]] = (),
) -> TimeStep:
```

If we simply passed in un-named arguments through the `cond` like done in other envs it would pass the extras as discounts to the `transition` branch. There isn't really a clean solution here unless JAX allows for named arguments to `jax.lax.cond`

**Comment by clement-bonnet** [comment]
Hi, thank you for your comment and your observation. Yes, we would rather have what you suggested, i.e. a simple `jax.lax.cond` statement. However, as far as I know, `cond` statements only take positional argument (as opposed to kwargs), and the presence of extras in `Connector` and `Game2048` that you mentioned above does not play well with the termination and transition functions. The transition function has an additional optional `discount` argument between observations and extras, which the termination function does not have. Therefore, they do not have the same signature as far as it concerns the first three positional arguments. So you can't simply use a `cond` statement.

**Comment by clement-bonnet** [comment]
Thank you @sash-a for replying so fast that I didn't see your comment! Agree with what you said.

A solution would be to move the `discount` argument from `transition` to after the `extras` argument so that they share the same `(reward, observation, extras)` positional arguments. Unless as you mentioned jax implements kwargs for cond.

**Comment by sash-a** [comment]
Ye that seems like a reasonable solution

**Comment by helpingstar** [comment]
It seems that the cases including the `lambda` form also included `extras`. Thank you both for your kind guidance. I also think @clement-bonnet 's solution looks good.

Since all instances calling `transition` use `extras` as keyword arguments, it seems possible to change the function definition. However, considering potential version issues, I’m cautious about submitting a PR right away.

I’m learning a lot from studying Jumanji’s code and can see the effort put into systematically building the JAX reinforcement learning environment throughout. Thank you very much for writing such excellent code.

I hope @clement-bonnet 's solution is implemented (or that `jax.lax.cond` will support keyword arguments). I'll leave the issue open. Feel free to close it at any time!

**Comment by sash-a** [comment]
@helpingstar would you be interested in making a PR to fix this?

**Comment by helpingstar** [comment]
@sash-a I’d be glad to help with this. Is there a preferred timeline?

**Comment by sash-a** [comment]
No rush honestly, whenever you have time! I think @clement-bonnet suggested fix should work well

**Comment by helpingstar** [comment]
@sash-a  Got it, thanks for letting me know!
